authz/authorize.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2025 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authz
+
+import (
+	"fmt"
+	"slices"
+)
+
+// Authorizer is the primary interface in order to check if a principal is granted
+// a given permission.
+type Authorizer interface {
+	// Check returns true if the principal has permission for the given action.
+	Check(p PrincipalID) bool
+}
+
+// CompileAuthorizer creates an authorizer for a Policy for a single resource.
+//
+// For example, if a resource `org/acme/dataplane/bar/mcpserver/qux`
+// wanted to check a `tool_invoke` permission, then it would do the following:
+//
+//	func RunMCPServer() error {
+//	  enforcer, err := authz.CompileAuthorizer(MustLoadAuthzPolicy(), "org/acme/dataplane/bar/mcpserver/qux", "tool_invoke")
+//	  if err != nil {
+//	  	return err
+//	  }
+//	  server.AddToolMiddleware(func(ctx context.Context) error {
+//	  	principal, err := CheckAuthentication(ctx)
+//	  	if err != nil {
+//	  		return fmt.Errorf("unauthorized", err)
+//	  	}
+//	  	if !enforcer.Check(PrincipalID(principal)) {
```

Comment:
how can a service use this interface to check non-static IDs?

e.g. incoming request is for MCPServer `abc`. but another request is for `def`. this example seems to be static, e.g. the entire for is for `abc` only.

this is not needed in current MCPServer or AiAgent, but in e.g. the management API (CRUDL) it is required, a caller can work with many different resources.

Reply:
Okay I updated the API to have a single level of SubResource for checking.